File: 383.ransom-note.cpp
#include <bits/stdc++.h>
using namespace std;
/*
 * @lc app=leetcode id=383 lang=cpp
 *
 * [383] Ransom Note
 */

// @lc code=start
class Solution
{
public:
    bool canConstruct(string ransomNote, string magazine)
    {
        if (magazine.size() < ransomNote.size())
        {
            return false;
        }

        unordered_map<char, int> ump;
        for (auto &letter : magazine)
        {
            ump[letter]++;
        }

        for (auto &letter : ransomNote)
        {
            if (ump[letter] == 0)
            {
                return false;
            }
            ump[letter]--;
        }
        return true;
    }
};
// @lc code=end

File: Java/Grind75/Week2/383.ransom-note.java
package Grind75.Week2;

import java.util.*;
/*
 * @lc app=leetcode id=383 lang=java
 *
 * [383] Ransom Note
 */

// @lc code=start
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        Map<Character, Integer> freq = new HashMap<>();
        for (char ch : magazine.toCharArray()) {
            freq.put(ch, freq.getOrDefault(ch, 0) + 1);
        }
        for (char ch : ransomNote.toCharArray()) {
            if ((freq.getOrDefault(ch, 0) - 1) < 0)
                return false;
            freq.put(ch, freq.get(ch) - 1);
        }
        return true;
    }
}
// @lc code=end